### ISSUE TITLE
"Family" of objects and opt-in

### ISSUE THREAD

**Issue by petronius12** [google/physical-web#593]
Hi! We are very excited about this project.
Is it possible to create a "family" of objects , where the user choice to tap one object once lets automatically opt-in for all the objects of the same "family" (such as those of the same domain)?

Thanks!

**Comment by scottjenson** [comment]
Interesting idea. Could you clarify a bit for me? Once the user 'opts in' to beacons from one domain, how would the user experience change?

**Comment by petronius12** [comment]
Hi Scott! Thanks for your answer.
I'll explain my idea through a very simple example: I am a student and I always catch the bus at stop A. So I did my first opt-in at stop A (bus ETA, traffic info, etc.). If I am around stop B (same "beacon family/domain" of stop A) it would be useful to have those informations on my screen without "opting in" at stop 2.
In other words: you allow, opting-in once, all the same-domain-beacons to exit the proxy system.

**Comment by roalfred** [comment]
This is interesting because I am exploring using the beacons in the same way, for providing transit information on public transport

Sent from my iPhone

On 17 Feb 2016, at 3:17 PM, petronius12 <notifications@github.com<mailto:notifications@github.com>> wrote:


Hi Scott! Thanks for your answer.
I'll explain my idea through a very simple example: I am a student and I always catch the bus at stop A. So I did my first opt-in at stop A (bus ETA, traffic info, etc.). If I am around stop B (same "beacon family/domain" of stop A) it would be useful to have those informations on my screen without "opting in" at stop 2.
In other words: you allow, opting-in once, all the same-domain-beacons to exit the proxy system.

-

**Comment by scottjenson** [comment]
Sorry, but what does 'exit the proxy system' really mean? It is very likely  the bus stop B web page has a different description (e.g. "Bus line B, heads toward city center") If we opted out of the proxy system, it's not clear what we would show the user.

**Comment by petronius12** [comment]
Ok, I'll try to explain better my idea.
Question A: An owner of a small business has a beacon transmitting an URL of offers / promo codes installed in his shop. A customer X chooses to "opt-in" in that webpage. Two days later, the same customer X come back in the shop A without opting-in in the webpage, but keeping active the Chrome functionality to find beacons around him. So he can see the webpage, but he doesn't need to opt-in again.
Can the owner of the shop know that the people who opted-in in the webpage come back in the shop (without another opt-in by the customer/user), thanks to the "call" by the customer X device to the owner server?
Question B: Same as above, but the owner has 2 shops: A and B. **Same domain** webpages (eg. grocery.com/shopA and grocery.com/shopB). Can the owner of the shops A&B know that the people who opted-in in the webpage at shop A come back, even if they come back in shop B (without another opt-in by the customer/user)?
Everything is anonymous, of course: no log-in needed, just a very useful data analysis.

**Comment by scottjenson** [comment]
At this point, skipping a user opt-in isn't possible. As the Physical Web is a universal scanner it could be used by any company to track users in any store so that is why we are requiring users to a) choose to see the list of nearby beacons and b) click on one each time they are nearby. This ensures the user really wants to interact with your web page. 

Your proposal is that anytime the user visits a store web page, that is adequate opt in to allow tracking in future stores. Unfortunately, this is simply impractical to do as this type of background scanning would require large amounts of battery. More importantly, it just doesn't feel like adequate opt-in by the user. We realize this limits your ability to track users and appreciate that this is a use case many companies would like to do. However, as this is a new technology, we feel it is exceptionally important to protect the user (possibly over protect at this early stage) so they can trust in how the product works. 

However, a variation of this task should eventually be possible using service worker as that would require a clear and obvious user opt-in. So this idea could happen, with adequate user protections, in the future. It's just not possible right now.

Scott

**Comment by petronius12** [comment]
Thank you for your answer. We have 2 more questions:
- is it possible for a developer to change the beacon URL description "dynamically"? I mean, is it possible to write command-lines in order to change that description according to different parameters?
- is this change "real-time" or does it need some time to refresh?

**Comment by scottjenson** [comment]
Sorry for the delay. We've been approached by others on this question. This really depends on the caching that is done by the scanner. Our scanner holds information for at least 5 minutes before rescanning the page on the next request. Other scanners (like Opera's) may be different.

This means that if you update the description in the page:

       <meta name="description" content="something">

"something" will show up in the list view in approximately 5 minutes. There is a proposal to let the web page suggest how long caching should take place by just looking at the cache headers. This is under consideration but not finalized yet.

Does this answer your question?

**Comment by tolson2000** [comment]
@petronius12 Both the meta description and the title tag can dynamically change based on the complete URL received. There is a early example on the youtube of the weather wear beacon changing the Title and Description based on the complete URL. It is easy to accomplish on the serverside with the likes of PHP, etal. The original example changed often by cycling through temperature, pressure, and humidity. In that case a cacheing of 5 minutes is more than reasonable for that app. The real time comes from the fact that the complete URL is changing.

**Comment by petronius12** [comment]
Yes, it answered my question. Thank you so much for your time, you have been very helpful to us.
Any idea about the Chrome 49/mobile versions release date?

**Comment by scottjenson** [comment]
It should be rolling out over the next few weeks

**Comment by mmocny** [comment]
Expanding on Scott's suggestion to use Service Worker:

- Physical Web opt-in is universal (you only need to do it once, for all urls), but all it does it let you see nearby results.  It will *never* automatically open pages or give the page any signal that you are nearby.

- Once you *open* a URL, there are ways to have that page persist on your phone and continue to run in the background.  E.g. add-to-homescreen / service-worker.  This can be "per-domain", and is not tied to physical web / beacons in any way.  It's just a thing the web can do these days.

- Unfortunately, while the web does have access to geolocation, there are not yet web api's for geofencing or beacon scanning (which you would need to accomplish the task you wanted).  These would be useful apis, but they simply aren't available today.

So, I think for today, you would have to write a native app to solve your problem -- but then you lose the benefits of discovery with Physical Web.

(Also, it's not necessarily a good idea to track users like this, you should be clear what you intend and tread carefully!)
